qutip/core/numpy_backend.py

```diff
@@ -2,8 +2,46 @@
 
 
 class NumpyBackend:
+    __slots__ = (
+        "_qt_np",
+        "abs",
+        "angle",
+        "arccos",
+        "array",
+        "ceil",
+        "conj",
+        "cos",
+        "cumsum",
+        "diff",
+        "e",
+        "histogram",
+        "imag",
+        "inf",
+        "inner",
+        "linalg",
+        "log",
+        "log2",
+        "max",
+        "maximum",
+        "ones",
+        "pi",
+        "prod",
+        "real",
+        "searchsorted",
+        "sin",
+        "sort",
+        "sqrt",
+        "sum",
+        "where",
+        "zeros",
+        "zeros_like",
+    )
+
     def _qutip_setting_backend(self, np):
         self._qt_np = np
+        for slot in self.__slots__:
+            if slot != "_qt_np":
+                setattr(self, slot, getattr(np, slot))
 
     def __getattr__(self, name):
```

Comment:
Is the `__getattr__` still needed? Is it to allow access to attributes that are not in `__slots__`? If so, do we actually want to allow arbitrary attribute access on the backend? Or only the ones we specify here?

Reply:
It will help external contributors, ideally they should not need to know how this work to contribute.
I also fear that some call in a line that is missed in tests will creep up eventually and with this it will still work.